module/input/SpeechIntent/src/SpeechIntent.cpp

```diff
@@ -0,0 +1,425 @@
+
+                std::unique_ptr<SpeechIntentMsg> intent = parse_voice2json_json(intent_json, strlen(intent_json));
+                emit(std::move(intent));
+
+                free(intent_json);
+            }
+        });
+
+        // Read any input from stderr and print it. We cannot proceed with normal operations so we just print
+        // the error messages
+        on<IO>(voice2json_proc.stderr, IO::READ).then([this] {
+            char buffer[0x1000];
+            ssize_t bytes_read = read(voice2json_proc.stderr, buffer, sizeof(buffer) - 1);
+            if (bytes_read == -1) {
+                NUClear::log<NUClear::FATAL>(
+                    fmt::format("({}:{}) Failed to read from stderr, errno = {}", __FILE__, __LINE__, errno));
+                return;
+            }
+            buffer[bytes_read] = 0;
+            NUClear::log<NUClear::FATAL>(fmt::format("({}:{})\nVOICE2JSON stderr, errno = {} ({} bytes)\n{}",
+                                                     __FILE__,
+                                                     __LINE__,
+                                                     errno,
+                                                     bytes_read,
+                                                     buffer));
+        });
+
+        // This trigger allows the emission of a "SpeechIntentMsg" to be enabled/disabled.
+        on<Trigger<SpeechInputSetOutputMsg>>().then(
+            [this](const SpeechInputSetOutputMsg& msg) { output_enabled = msg.enabled; });
+
+        // This trigger will recognize the intent of a wav file.
+        on<Trigger<SpeechInputRecognizeWavFile>>().then([this](const SpeechInputRecognizeWavFile& msg) {
+            if (!write_audio_to_file(voice2json_proc.stdin, (char*) msg.filename.data())) {
+                NUClear::log<NUClear::FATAL>(
+                    fmt::format("({}:{}) Failed to write to {}", __FILE__, __LINE__, msg.filename.data()));
+            }
+        });
+    }
+
+    // Convenience function for sending a "SpeechInputRecognizeWavFile" message to this module
+    void SpeechIntent::recognize_wav(std::string wav_filename) {
+        std::unique_ptr<SpeechInputRecognizeWavFile> msg =
+            std::make_unique<SpeechInputRecognizeWavFile>(SpeechInputRecognizeWavFile{wav_filename});
+        emit(std::move(msg));
+    }
+
+    SpeechIntent::SpeechIntent(std::unique_ptr<NUClear::Environment> environment) : Reactor(std::move(environment)) {
+
+        on<Configuration>("SpeechIntent.yaml").then([this](const Configuration& cfg) {
+            this->log_level = cfg["log_level"].as<NUClear::LogLevel>();
+        });
+
+        // Choose between 3 modes based on command line arguments. The "cli" and "file" arguments are for testing
+        // purposes. In the case where no cmd line arguments are passed the module is initialized in
+        // "TRANSCRIBE_MODE_STREAM" mode.
+        on<Trigger<CommandLineArguments>>().then([this](const CommandLineArguments& args) {
```

Comment:
@Bidski, @tayloryoung6396, @c3144856 This was the section that we were discussing in the previous pull request.

Reply:
> Its pointless because you can achieve all of the same functionality using the configuration file. You could easily set up a `sound_source` entry in the config file, and if that represents a valid file (you can use `std::filesystem` for this) then you load that as an audio file, if `sound_source` is `mic` (or some other appropriate keyword) then you look for the appropriate hardware and get your audio data from there.
> 

The perspective that I was thinking about this from was that hard coding the path to the file means that you'd need to edit the file in order to change which sentences is processed by the module, and that could accidentally make its way into version control.

I'm happy to change this, though I think it will make using it a little bit more awkward when verifying the command works.

> Also, are you aware of [this](https://stackoverflow.com/a/48795232/4795763)? Not sure if this would work in WSL.

Yes. `--device` doesn't work on macOS or Windows because docker desktop runs inside a virtualised environment which doesn't have access to the device drivers. [See here](https://github.com/docker/for-mac/issues/900).

Reply:
> The perspective that I was thinking about this from was that hard coding the path to the file means that you'd need to edit the file in order to change which sentences is processed by the module, and that could accidentally make its way into version control.

This isn't really much of an issue. The per-robot/per-binary config system will probably be your friend here. Given that we will never be able to use the microphone in docker (at least not in a cross-platform way) so "hard coding" a path in the `docker` "robot" config file is not a horrible idea.

I would make a data folder `module/input/SpeechIntent/data/audio` and stick some test WAV files in there, your "hard coded" path would then just be `audio/test1.wav`.

Also consider unit tests will not have access to a microphone either, so if you set things up properly you will be keeping some data files in the repo anyway.